Open Telemetry

sdelamo opened the pull request.

sdelamo committed:
open tracing features

sdelamo commented:
Waiting for Tracing release with Open Telemetry

sdelamo committed:
add aws sdk instrumentation open telemetry dependencies

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryAnnotations.java:
```diff
@@ -0,0 +1,54 @@
+import jakarta.inject.Singleton;
+
+@Singleton
+public class OpenTelemetryAnnotations implements OpenTelemetryFeature {
```
Don't think the annotation processor is needed.

What is needed is implementation("io.micronaut.tracing:micronaut-tracing-opentelemetry"). It brings in the opentelemetry-annotations.

sdelamo replied:
According to our docs, I think the annotation processor is needed for the mappers. 

https://micronaut-projects.github.io/micronaut-tracing/latest/guide/#annotations

shingjo replied:
Possibly, the test project I used to create the GCP guide doesn't need it to build or run. Am using Java 17 and it might be needed with older versions.

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryExporterFeature.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    default void apply(GeneratorContext generatorContext) {
+        generatorContext.addDependency(OpenTelemetryDependencyUtils.openTelemetryDependency()
```
Applies to all Exporters:

Does not add the dependency "io.micronaut.tracing:micronaut-tracing-opentelemetry". 
Granted, doesn't fail the build, but doesn't add the annotation processor nor the AWS X-Ray ID Generator code.

sdelamo replied:
I did this commit https://github.com/micronaut-projects/micronaut-starter/commit/d402e7e65b21124e162f1a166beb002c30d508ba

Does it solve this issue?

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryExporterFeature.java:
```diff
@@ -0,0 +1,63 @@
+        generatorContext.addDependency(OpenTelemetryDependencyUtils.openTelemetryDependency()
+                .artifactId(exporterArtifactId())
+                .compile());
+        generatorContext.getConfiguration().put("otel.traces.exporter", exporterName().toLowerCase(Locale.ROOT));
```
Applies to all Exporters:

FYI: Might want to use `generatorContext.getConfiguration().addNested`, since there can be multiple exporters. Not sure if starter will ever configure more than one, but future proofing.

sdelamo replied:
see https://github.com/micronaut-projects/micronaut-starter/pull/1295/commits/c22db22860c262bdb9a26004c7df1d0cdef81697

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryHttp.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public void apply(GeneratorContext generatorContext) {
```
This brings in "io.micronaut.tracing:micronaut-tracing-opentelemetry", but we might want to add it as a dependency that way they can remove the http dependency if they don't want to trace the "http" calls.

sdelamo replied:
I did this commit https://github.com/micronaut-projects/micronaut-starter/pull/1295/commits/d402e7e65b21124e162f1a166beb002c30d508ba

Does it solve this issue?

shingjo reviewed (commented):
We also discussed adding the "micronaut-tracing-opentelemetry-http" dependency if the application type was 'default', so that it was constant with the current OpenTracing implementation.

sdelamo committed:
Merge branch '3.6.x' into opentracing

sdelamo committed:
remove unused import

sdelamo committed:
use addNested

sdelamo committed:
apply openTelemetry if not http for exporters

sdelamo committed:
remove unused imports

sdelamo committed:
add Micronaut OTEL gRPC

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryExporterGoogleCloudTrace.java:
```diff
@@ -0,0 +1,53 @@
+import jakarta.inject.Singleton;
+
+@Singleton
+public class OpenTelemetryExporterGoogleCloudTrace extends OpenTelemetryExporterFeature {
```
NOTE: If someone only wanted to use GCP tracing and GraalVM, this doesn't add the required GraalVM build dependency (see: public abstract class GcpFeature implements Feature).

sdelamo replied:
I think this should solve this: https://github.com/micronaut-projects/micronaut-starter/pull/1295/commits/bd982c263cc40e9b1c4465c7486e174a8b01a166

shingjo commented on starter-core/src/main/java/io/micronaut/starter/feature/opentelemetry/OpenTelemetryExporter.java:
```diff
@@ -0,0 +1,25 @@
+
+import io.micronaut.starter.feature.OneOfFeature;
+
+public interface OpenTelemetryExporter extends OneOfFeature {
```
The `OneOfFeature` doesn't allow you to specify the `logging` exporter in addition to `zipkin`|`jaeger`|`gcp`, which OpenTelemetry supports. (Granted the `logging` exporter is primarily used for debugging)

sdelamo replied:
I did not know you could/want to use multiple exporter at the same time. I will remove the `OneOfFeature`.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-starter/pull/1295/commits/5c67ac3f741f3d41477292f0b9f4a0be7aacdfe4

sdelamo committed:
Merge branch '3.6.x' into opentracing

sdelamo committed:
delete OpenTelemetryExporter

A user may want multiple exporters in his application

sdelamo committed:
apply profile and dependency for gcp exporter

sdelamo committed:
simplify otel features

There will be only five Otel features:

- `tracing-opentelemetry-zipkin`
- `tracing-opentelemetry-jaeger`
- `tracing-opentelemetry-aws`
- `tracing-opentelemetry-gcp`
- `tracing-opentelemetry-exporter-logging`

sdelamo commented:
I have simplified so that there are  only five visible Otel features: 

- `tracing-opentelemetry-zipkin`
- `tracing-opentelemetry-jaeger`
- `tracing-opentelemetry-aws`
- `tracing-opentelemetry-gcp`
- `tracing-opentelemetry-exporter-logging`

sdelamo committed:
fix: multiple exporter configuration

sdelamo committed:
checkstyle

sdelamo committed:
make otlp experter visible

sdelamo committed:
List of exporters to be used for tracing, separated by commas

see: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#exporters

Commit:
Change format to nested

sdelamo committed:
exclusions is otel.exclusions
